### PR TITLE
Improved tests for scan partition prune-able PMS [HZ-2854]

### PR DESCRIPTION
**Pull request by Fly-Style**
Based on #25116

### CODE REVIEW

**Review by ivanthescientist — APPROVED** [review]

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java [hunk 1]
```diff
@@ -22,110 +22,233 @@
...
     public static void beforeClass() throws Exception {
-        initialize(5, null);
+        initialize(MEMBERS, null);
     }
 
     @Before
     public void before() throws Exception {
+        partitionAssignment = getPartitionAssignment(instance());
+        reversedPartitionAssignment = new HashMap<>();
+        for (Map.Entry<Address, int[]> entry : partitionAssignment.entrySet()) {
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:94]: should be easier using Java streams

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java [hunk 2]
```diff
@@ -22,110 +22,233 @@
...
+                perMemberOwnedPId[j++] = pIdCandidate;
+                clusterAddresses.remove(address);
+            }
+        }
     }
 
     // Note: basic functionality must work even without member pruning used.
     @Test
     public void test_basic() {
         // Basic test is performed as submitting job by DAG.
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:145]: what is basic test supposed to test? before it used scan pruning, now it does not. Rename the method to be less confusing.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java [hunk 3]
```diff
@@ -22,110 +22,233 @@
...
+        Vertex sourceV = dag.getVertex("source");
+        assertNotNull(sourceV);
+
+        ProcessorMetaSupplier metaSupplier = sourceV.getMetaSupplier();
+        assertNotNull(metaSupplier);
+        assertInstanceOf(SpecificPartitionsImapReaderPms.class, metaSupplier);
+
+        SpecificPartitionsImapReaderPms pms = (SpecificPartitionsImapReaderPms) metaSupplier;
+        // Ensure we scan all partitions.
+        assertNull(pms.partitionsToScan);
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:167]: ??? is that unit test of `mapReader` static method? If so, it should be separate. PMS is immutable in DAG - do you expect that something might have changed it?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java [hunk 4]
```diff
@@ -22,110 +22,233 @@
...
+        instance().getJet().newLightJob(dag, jobConfig).join();
 
         // Then
-        Assert.assertEquals(1, ((SpecificPartitionsImapReaderPms) pms).partitionsToScan.length);
-        Assert.assertEquals(keyOwnerPartitionId, ((SpecificPartitionsImapReaderPms) pms).partitionsToScan[0]);
+        assertPrunability(partitionsToUse);
+    }
+
+    private DAG setupPrunableDag(IMap<Integer, Integer> map, int partitionCountToUse) {
+        assertGreaterOrEquals("Should scan at least one partition", partitionCountToUse, 1);
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:198]: is empty partitions list invalid for SpecificPartitionsImapReaderPms? There is no such validation and it should work just fine in that case.
We should test edge cases.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java (hunk 3 above)
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:167]: ok, you rely on the fact that DAG is not serialized when submitting light job (which is only optmiziation implementation detail). Still, this feels more like unit test. Mixing layers and depending on implementation deatls makes the higher level test less readable.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java [hunk 5]
```diff
@@ -22,110 +22,233 @@
...
+            }
+        }
+
         mapName = randomName();
         sinkName = randomName();
+        sourceMap = instance().getMap(mapName);
+        sinkMap = instance().getMap(sinkName);
 
+        // By this, we ensure that PMS works without member pruning being used.
         jobConfig = new JobConfig().setArgument(SQL_ARGUMENTS_KEY_NAME, emptyList());
```
[the remark this answers is not in the document]
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:106]: the comment is misleading  "By this" = "by adding empty SQL_ARGUMENTS_KEY_NAME", which is not true.

But I noticed another problem. After the changes from https://github.com/hazelcast/hazelcast/commit/00745ffa01c6c2af72d082dec1c0741ab1a93444#diff-d4ae1034ef6cf9f4be39bf282a7bba5e5f0517c3167c055a47af9448a497c27a, PMSes use different serialization service for expression evaluation than PS and P. It is not even serialization service for the member, but service with default config. Also those changes accidenatally do not require SQL_ARGUMENTS_KEY_NAME for PMS expression eval context.

[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:106]: > Also after those changes we _accidentally_ do not require SQL_ARGUMENTS_KEY_NAME for PMS expression eval context.

It's not 'accidentally', it was the only reason this change: previously, we had no need for EEC for PMS -- it was done ONLY for the new prunable IMap scan PMS, where we're evaluating expressions for the first time in PMS.

[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:106]: > what should be the correct serialization service for PMS?

Default is correct.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java (hunk 4 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:198]: That method is used when we are setting up a case where prunability **must** work. You may notice in `test_basic`, that it tests scanning for basic scenario without declared partitioned strategy.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java (hunk 3 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:167]: > Mixing layers and depending on implementation details makes the higher level test less readable.

I do not mix anything here, I am working on `CoreAPI` to have access to vertexes, PMS-es to check all I need.
Again, tt was done **specially** on CoreAPI level with purpose. E.g, take a look on `WriteFilePTest`

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java (hunk 1 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:94]: Taste of choice, IMO.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java]: In the linked commit we used `DefaultSerializationServiceBuilder().build()` only for  `TestProcessorMetaSupplierContext`, so only in tests (likely ones that did not start Hazelcast instance at all). This does not seem like a correct default for production use, in particular:

> It is not even serialization service for the member

So you do not use any serializers configured in the member. What if the query parameter has custom serializer defined?

I was wondering also if query parameter (or other expression evaluated by any PMS) can use class upload with the job. Obviously this is not possible in SQL (you cannot attach class even to `CREATE JOB`) but PMS can be also used directly. Also users can define their own PMSes.

[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java]: > So you do not use any serializers configured in the member. What if the query parameter has custom serializer defined?

Sorry, I don't see any problem here. PMS-ses are `Serializable`, the only reason to use `ExpressionEvalContext` is to evaluate expressions. In fact, in a new PMS for map scan, we're using `EEC` to evaluate constant expressions, we do not evaluate them on rows.

Moreover, that's what I was afraid of making a way with expression evaluations inside PMS - we faced an issue that required revamping the existing design for PMS-es : they are not aware of the serializer by design. 

So, my vote is still the same: default serialization service suits well to evaluate constant expressions in `SpecificPartitionsImapReaderPms`.

**Comment by Fly-Style** [comment]
run-lab-run

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java (hunk 1 above)
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java:84]: with some static imports:
```java
        reversedPartitionAssignment = partitionAssignment.entrySet().stream()
                .flatMap(entry -> stream(entry.getValue()).mapToObj(pId -> entry(pId, entry.getKey())))
                .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
```
not that simple as I hoped

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java (hunk 3 above)
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java]: you mix integration test with unit test. I meant this (method names are not perfect): https://github.com/hazelcast/hazelcast/pull/25206/commits/32f13586ee34c8c89ffb8cd80d5f25bede4c2c8f

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java (hunk 4 above)
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java]: unit tests should check also boundary conditions. Added such check.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java (hunk 3 above)
Comment (k-jamroz) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SpecificPartitionsImapReaderPmsTest.java]: it would be nice to even even more unit tests, the algorithm in SpecificPartitionsImapReaderPms is quite complicated and not all cases and branches are covered, we also lack unit tests with non-null strategy

**Review by k-jamroz — APPROVED** [review]
